File: leetcode-cn/1817.0_Finding_the_Users_Active_Minutes.py
'''
hash table

执行用时：180 ms, 在所有 Python3 提交中击败了15.66% 的用户
内存消耗：20.6 MB, 在所有 Python3 提交中击败了49.40% 的用户
通过测试用例：38 / 38
'''
class Solution:
    def findingUsersActiveMinutes(self, logs: List[List[int]], k: int) -> List[int]:
        id2times = defaultdict(set)
        for id, time in logs:
            id2times[id].add(time)
        freq = Counter(len(v) for v in id2times.values())
        return [freq[i] for i in range(1, k + 1)]

'''
hash table

执行用时：100 ms, 在所有 Python3 提交中击败了84.34% 的用户
内存消耗：20.9 MB, 在所有 Python3 提交中击败了36.15% 的用户
通过测试用例：38 / 38
'''
class Solution:
    def findingUsersActiveMinutes(self, logs: List[List[int]], k: int) -> List[int]:
        id2times = defaultdict(set)
        for id, time in logs:
            id2times[id].add(time)
        ans = [0] * k
        for id, times in id2times.items():
            ans[len(times) - 1] += 1
        return ans 

File: go/1817.0_Finding_the_Users_Active_Minutes.go
/*
hash table

执行用时：124 ms, 在所有 Go 提交中击败了61.54% 的用户
内存消耗：9.3 MB, 在所有 Go 提交中击败了30.77% 的用户
通过测试用例：38 / 38
*/
func findingUsersActiveMinutes(logs [][]int, k int) []int {
    id2times := map[int]map[int]int{}
    for i := 0; i < len(logs); i++ {
        if id2times[logs[i][0]] == nil {
            id2times[logs[i][0]] = map[int]int{}
        }
        id2times[logs[i][0]][logs[i][1]]++
    }
    ans := make([]int, k)
    for _, v := range id2times {
        if len(v) - 1 < 0 {
            continue
        }
        ans[len(v) - 1]++
    }
    return ans 
}
